### DISCUSSION TITLE
[JS Hooks] Is there any method to get all records, without filter or any specific criteria?

### DISCUSSION THREAD

**Discussion by JunoNgx** [pocketbase/pocketbase#4818] [Q&A] [answered]
Context: I'm implementing a soft-deleting feature, and performing the related migration. Each user needs to have a corresponding record in the `trashBins` collections. I'm retrieving all the records in the `users` table and `trashBins` table and create a new `trashBins` record for the user if it has not existed yet.

I've been looking for a really long time at the [documentation](https://pocketbase.io/docs/js-records/#fetch-multiple-records), but the three provided options for fetching multiple records all come with kinds of filters:
```
// retrieve multiple "articles" collection records by their ids
const records = $app.dao().findRecordsByIds("articles", ["RECORD_ID1", "RECORD_ID2"])

// retrieve multiple "articles" collection records by a custom dbx expression(s)
// (for all avalaible expressions, please check the Database guide)
const records = $app.dao().findRecordsByExpr("articles",
    $dbx.exp("LOWER(username) = {:username}", { "username": "John.Doe" }),
    $dbx.hashExp({ status: "pending" })
)

// retrieve multiple "articles" collection records by a string filter expression
// (use "{:placeholder}" to safely bind untrusted user input parameters)
const records = $app.dao().findRecordsByFilter(
    "articles",                                    // collection
    "status = 'public' && category = {:category}", // filter
    "-publised",                                   // sort
    10,                                            // limit
    0,                                             // offset
    { category: "news" },                          // optional filter params
)
```

Is there something along the line of `dao.getAllRecords("users")`?

**Comment by ganigeorgiev** [comment]
You can try:

```js
const records = arrayOf(new Record);
$app.dao().recordQuery("users").all(records)

// or call with empty expressions (in the next release this is renamed to findAllRecords)
$app.dao().findRecordsByExpr("users")
```

**Reply by JunoNgx** [reply]
I'm performing this in a migration script:

```
/// <reference path="../pb_data/types.d.ts" />

migrate((db) => {
    const dao = new Dao(db);

    const getAllRecords = (collectionName) => {
        const records = arrayOf(new Record);

        dao.recordQuery(collectionName)
            .all(records);

        return records;
    };

    const allUsers = getAllRecords("users");

    for (const user of allUsers) {
        // my logic
    };
}, (db) => {
    // Intentionally does nothing
})

```

The error message I'm receiving for this is `Error: Failed to apply migration 1713928508_create_trashbins_for_users.js: sql: no rows in result set`.

Or is there a different way to access the `dao` instance from here?

**Reply by JunoNgx** [reply]
Using `dao.findRecordsByExpr("users");` still gets me `sql: no rows in result set` though I'm pretty sure that I have two users in the `users` table.

**Reply by ganigeorgiev** [reply, marked as answer]
I'm not able to reproduce it. Even without records the `all()` method should NOT throw this error and return an empty array as long as the collection name is valid.

There is something else in your migration file that triggers the above error but the above code doesn't show it.

**Reply by JunoNgx** [reply]
You're correct. The error is from inside my logic to create the `trashBins` record.

I have another question: how do I manually clear the internal table `_migrations` to prompt the logic to go through everything in `pb_migrations` again?  I have been renaming the file to get migration logic to be executed again, but this is clearly not the right way.

I found `./pocketbase migrate history-sync`, but this is not exactly what I'm looking for either.

Is there anything start fresh and get the migration scripts to be executed again?

**Reply by ganigeorgiev** [reply]
> Is there anything start fresh and get the migration scripts to be executed again?

No, because it could fail for example if you are creating an index or table with a raw DB expression.

Migrations are executed and reverted in their specific order. If you want to start from "fresh" without any data, you can delete your `pb_data` and/or run `migrate down [n]` to rollback "n" number of migrations.

If you have other questions please file a new Q&A discussion with more details about your use case and what you are trying to do.

**Reply by JunoNgx** [reply]
Thank you very much for your help, `migrate down [n]` is exactly what I've been looking for.
